contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,363 @@
+        uint256 qy
+    ) internal view returns (bool valid, bool supported) {
+        (bool success, bytes memory returndata) = address(0x100).staticcall(abi.encode(h, r, s, qx, qy));
+        return (success && returndata.length == 0x20) ? (abi.decode(returndata, (bool)), true) : (false, false);
+    }
+
+    /**
+     * @dev signature verification - solidity implementation. This version will work on all chains, but is more
+     * expensive to use, and to deploy, than the version that rely on precompiles.
+     *
+     * @param h - hashed message
+     * @param r - signature half R
+     * @param s - signature half S
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     */
+    function verifySolidity(uint256 h, uint256 r, uint256 s, uint256 qx, uint256 qy) internal view returns (bool) {
+        if (r == 0 || r >= N || s == 0 || s >= N || !isOnCurve(qx, qy)) return false;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(qx, qy);
+        uint256 w = Math.invModPrime(s, N);
+        uint256 u1 = mulmod(h, w, N);
+        uint256 u2 = mulmod(r, w, N);
+        (uint256 x, ) = _jMultShamir(points, u1, u2);
+        return (x == r);
+    }
+
+    /**
+     * @dev public key recovery
+     * @param h - hashed message
+     * @param v - signature recovery param
+     * @param r - signature half R
+     * @param s - signature half S
+     */
+    function recovery(uint256 h, uint8 v, uint256 r, uint256 s) internal view returns (uint256, uint256) {
+        if (r == 0 || r >= N || s == 0 || s >= N || v > 1) return (0, 0);
+
+        uint256 rx = r;
+        uint256 ry2 = addmod(mulmod(addmod(mulmod(rx, rx, P), A, P), rx, P), B, P); // weierstrass equation y² = x³ + a.x + b
+        uint256 ry = Math.modExp(ry2, P1DIV4, P); // This formula for sqrt work because P ≡ 3 (mod 4)
+        if (mulmod(ry, ry, P) != ry2) return (0, 0); // Sanity check
+        if (ry % 2 != v % 2) ry = P - ry;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(rx, ry);
+        uint256 w = Math.invModPrime(r, N);
+        uint256 u1 = mulmod(N - (h % N), w, N);
+        uint256 u2 = mulmod(s, w, N);
+        (uint256 x, uint256 y) = _jMultShamir(points, u1, u2);
+        return (x, y);
+    }
+
+    /**
+     * @dev address recovery
+     * @param h - hashed message
+     * @param v - signature recovery param
+     * @param r - signature half R
+     * @param s - signature half S
+     */
+    function recoveryAddress(uint256 h, uint8 v, uint256 r, uint256 s) internal view returns (address) {
+        (uint256 qx, uint256 qy) = recovery(h, v, r, s);
+        return getAddress(qx, qy);
+    }
+
+    /**
+     * @dev derivate public key
+     * @param privateKey - private key
+     */
+    function getPublicKey(uint256 privateKey) internal view returns (uint256, uint256) {
+        (uint256 x, uint256 y, uint256 z) = _jMult(GX, GY, 1, privateKey);
+        return _affineFromJacobian(x, y, z);
+    }
```

Comment:
Is this function needed? I don't see a real use case for it as it would require exposing the private key anyway